src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs

```diff
@@ -373,19 +373,20 @@ public void ReadAsyncCompletesIfFlushAsyncCanceledMidFlush()
                     var readTask = Pipe.Reader.ReadAsync();
 
                     // Signal writer to initiate a flush
-                    if (!readTask.IsCompleted)
+                    if (!readTask.IsCompleted || readTask.Result.IsCompleted)
                     {
                         resetEvent.Set();
                     }
 
                     var result = await readTask;
-                    resetEvent.Reset();
 
                     if (result.Buffer.IsEmpty)
                     {
                         return;
                     }
 
+                    resetEvent.Reset();
```

Comment:
Can you help me understand why this re-ordering was necessary?

Reply:
So `resetEvent` is `Set` when reader completes so canceller doesn't block on `WaitOne`.

Reply:
> So `resetEvent` is `Set` when reader completes

Do you mean `Reset`? And why not move it to the very end of the loop (after the call to AdvanceTo)?

Reply:
No, I mean `Set`. So canceller can loop and exit.